pkg/snapshot/snapshotter/snapshotter.go

```diff
@@ -142,8 +143,11 @@ func (ssr *Snapshotter) stop() {
 		ssr.deltaSnapshotTimer.Stop()
 		ssr.deltaSnapshotTimer = nil
 	}
-	if ssr.cancelWatch != nil {
-		ssr.cancelWatch()
+	if ssr.etcdClient != nil {
```

Comment:
>What about cleaning up `ssr.watchCh`? This might be a potential memory leak.

Sorry for late reply. `ssr.WatchCh` is readonly, so can't close it. `etcdClient.Close()` call do the same internally.

> Might be a good idea to make this block into a function, say `closeEtcdClient()` or something. Then this can be called in the other places too.

`etcdClient.Close()` is single function call. So, don't feel of adding wrapper around it, just for additional `nil` check.

> Also, wouldn't `defer` be better here? Especially, for the `Stop()` calls to the times above.

Entire function above i.e. `(ssr *Snpashotter) Stop()` is called as `defer`. So,your suggestion is handled there.

Reply:
> `ssr.WatchCh` is readonly, so can't close it. `etcdClient.Close()` call do the same internally.

Sounds good.

> `etcdClient.Close()` is single function call. So, don't feel of adding wrapper around it, just for additional `nil` check.

The suggestion for the defer was not for the `nil` check. There were in fact two separate suggestions.

1. `defer` for the delegated `stop()` calls of `fullSnapshotTimer` and `deltaSnapshotTimer`. What if those fail and that causes the `etcdClient` not to be `close()`d?
1. Wrap the clean up of etcdClient in a function because it is being done in two [different](https://github.com/swapnilgm/etcd-backup-restore/blob/memory-prof-2/pkg/snapshot/snapshotter/snapshotter.go#L146) [places](https://github.com/swapnilgm/etcd-backup-restore/blob/memory-prof-2/pkg/snapshot/snapshotter/snapshotter.go#L189) in two different ways. Is there a real reason to handle the potential error [here](https://github.com/swapnilgm/etcd-backup-restore/blob/memory-prof-2/pkg/snapshot/snapshotter/snapshotter.go#L147) and not [here](https://github.com/swapnilgm/etcd-backup-restore/blob/memory-prof-2/pkg/snapshot/snapshotter/snapshotter.go#L190)?

Reply:
> 1. `defer` for the delegated `stop()` calls of `fullSnapshotTimer` and `deltaSnapshotTimer`. What if those fail and that causes the `etcdClient` not to be `close()`?

 `full/deltaSnapshotTimer.Stop()` guarantees to stop the timer without any error return. So, there won't be any error or code path will go to `etcdClient.Close()` call.

>2\. Wrap the clean up of `etcdClient` in a function because it is being done in two [different](https://github.com/swapnilgm/etcd-backup-restore/blob/memory-prof-2/pkg/snapshot/snapshotter/snapshotter.go#L146) [places](https://github.com/swapnilgm/etcd-backup-restore/blob/memory-prof-2/pkg/snapshot/snapshotter/snapshotter.go#L189) in two different ways. Is there a real reason to handle the potential error [here](https://github.com/swapnilgm/etcd-backup-restore/blob/memory-prof-2/pkg/snapshot/snapshotter/snapshotter.go#L147) and not [here](https://github.com/swapnilgm/etcd-backup-restore/blob/memory-prof-2/pkg/snapshot/snapshotter/snapshotter.go#L190)?

Ahh. Got it. Thanks. I will change that.

Reply:
> `full/deltaSnapshotTimer.Stop()` guarantees to stop the timer without any error return. So, there won't be any error. And code path will definitely go to `etcdClient.Close()` call.

Sounds good.